rxnetty/src/main/java/io/reactivex/netty/client/ConnectionPoolImpl.java

```diff
@@ -54,7 +58,13 @@
     private final RxClient.ServerInfo serverInfo;
     private final PoolConfig poolConfig;
     private final ScheduledExecutorService cleanupScheduler;
-    private final AtomicBoolean isShutdown = new AtomicBoolean();
+    private final AtomicInteger aquiredConnectionsCounter = new AtomicInteger();
+    //is needed to not accept any new connections that are aquired
+    private final AtomicBoolean isShutdownRequested = new AtomicBoolean();
+    //is needed to not perform the shutdown twice
+    private final AtomicBoolean isShutdownPerformed = new AtomicBoolean();
+    //is needed to not perform shutdown during the aquire of a connection
+    private final ReadWriteLock shutdownLock = new ReentrantReadWriteLock();
```

Comment:
So, I don't think we need to be absolutely sure that shutdown does not happen during acquire. There are the following reasons for it:

- There could always be already checked out connections which aren't too different from concurrent acquire and shutdown.
- This lock means that the `shutdown` can not proceed while acquires are happening. If the shutdown happens from the eventloop (it can for rejection of hosts), it blocks the eventloop.

I would suggest having a boolean guarded by `synchronized` which is the flag for shutdown. If `acquire` finds this flag to be `true` before acquire, it is rejected.

`release` should cleanup when `aquiredConnectionsCounter` is 0 and the pool is shutdown.

The above approach will have some connections, when `release` happens after `shutdown` but between `acquire` checking for `shutdown` and actually acquiring, that will be checkout even when the pool is cleaned up. I think we are ok with such behaviors if we check for `shutdown` within `onSubscribe` (which I have suggested below). WDYT?